extra/utils.py

```diff
@@ -209,12 +209,35 @@ def get_child(parent, key):
       obj = getattr(obj, k)
   return obj
 
-def _tree(lazydata):
-  if type(lazydata).__name__ == "LazyBuffer": return [f"━━ realized {lazydata.dtype.name} {lazydata.shape}"] if (lazydata.realized) else _tree(lazydata.op)
+from tinygrad.lazy import LazyBuffer, LazyOp
+from typing import Union
+
+ctr = 0
+circle_tracker:dict[int,tuple[int,any]] = {}
+
+def _tree(lazydata:Union[LazyBuffer, LazyOp]):
+  global ctr
+  ctr += 1
+  if type(lazydata) == LazyBuffer and not (lazydata.realized): lazydata = lazydata.op
+  if type(lazydata) == LazyBuffer : return [f"━━ realized {lazydata.dtype.name} {lazydata.shape} "]
+
+  if id(lazydata) in circle_tracker: return [f"━⬆︎ goto {circle_tracker[id(lazydata)][0]}: {lazydata.op.name}"]
+  circle_tracker[id(lazydata)] = (ctr, lazydata)
```

Comment:
Holding lazydata here we do not allow some buffers to be freed. Clear `circle_tracker` at the end of `get_tree`?

Reply:
okay thanks for the tip